ipcs/socket/socket.go

```diff
@@ -0,0 +1,197 @@
+// (c) 2020, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package socket
+
+import (
+	"encoding/binary"
+	"errors"
+	"io"
+	"net"
+	"os"
+	"sync"
+	"syscall"
+
+	"github.com/ava-labs/gecko/utils/logging"
+	"github.com/ava-labs/gecko/utils/wrappers"
+)
+
+var (
+	// ErrTimeout is returned when a socket operation times out
+	ErrTimeout = errors.New("timeout")
+)
+
+// Socket manages sending messages over a socket to many subscribed clients
+type Socket struct {
+	log logging.Logger
+
+	addr     string
+	accept   acceptFn
+	connLock *sync.RWMutex
+	conns    []net.Conn
+
+	quitCh chan struct{}
+	doneCh chan struct{}
+}
+
+// NewSocket creates a new socket object for the given address. It does not open
+// the socket until Listen is called.
+func NewSocket(addr string, log logging.Logger) *Socket {
+	return &Socket{
+		log: log,
+
+		addr:     addr,
+		accept:   accept,
+		connLock: &sync.RWMutex{},
+
+		quitCh: make(chan struct{}),
+		doneCh: make(chan struct{}),
+	}
+}
+
+// Listen starts listening on the socket for new connection
+func (s *Socket) Listen() error {
+	l, err := listen(s.addr)
+	if err != nil {
+		return err
+	}
+
+	// Start a loop that accepts new connections until told to quit
+	go func() {
+		for {
+			select {
+			case <-s.quitCh:
+				close(s.doneCh)
+				return
+			default:
+				s.accept(s, l)
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Send writes the given message to all connection clients
+func (s *Socket) Send(msg []byte) error {
+	// Prefix the message with an 8 byte length
+	lenBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(lenBytes, uint64(len(msg)))
+	msg = append(lenBytes, msg...)
+
+	// Get a copy of connections
+	s.connLock.RLock()
+	conns := make([]net.Conn, len(s.conns))
+	copy(conns, s.conns)
+	s.connLock.RUnlock()
+
+	// Write to each connection
+	var err error
+	for _, conn := range conns {
+		if _, err = conn.Write(msg); err != nil {
+			return err
```

Comment:
might be worth returning a little more context about the error here.

e.g. `return fmt.Errorf("failed to write message to %s: %w", conn.RemoteAddr(), err)